### PR TITLE
 enhancement: prune historical data

### PR DESCRIPTION
**Pull request by shiqizng**
## Summary
a replacement PR for #1237

### CODE REVIEW

File: exporters/postgresql/postgresql_exporter.go [hunk 1]
```diff
@@ -67,10 +75,21 @@ func (exp *postgresqlExporter) Init(_ context.Context, cfg plugins.PluginConfig,
...
+		exp.round = 0
+	} else {
+		return fmt.Errorf("Init() err getting next round: %v", err)
 	}
-	return err
+	// if data pruning is enabled
+	if !exp.cfg.Test && exp.cfg.Delete.Rounds > 0 {
+		exp.dm = util.MakeDataManager(exp.ctx, &exp.cfg.Delete, exp.db, logger)
+		exp.wg.Add(1)
+		go exp.dm.Delete(&exp.wg, &exp.round)
```
Comment (winder) [inline exporters/postgresql/postgresql_exporter.go:90]: Could this work?
```suggestion
		exp.wg.Add(1)
		go func(){
			defer exp.wg.Done()
			exp.dm.Delete(&exp.round)
		}
```

File: exporters/util/prune.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		logger:   logger,
+		ctx:      ctx,
+		duration: d,
+	}
+
+	return dm
+}
+
+// Delete removes data from the txn table in Postgres DB
+func (p *postgresql) Delete(wg *sync.WaitGroup, nextRound *uint64) {
```
Comment (winder) [inline exporters/util/prune.go:58]: Somehow indicate that this is the main service loop
```suggestion
func (p *postgresql) DeleteLoop(wg *sync.WaitGroup, nextRound *uint64) {
```

File: exporters/util/prune.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+					err := p.db.DeleteTransactions(p.ctx, keep)
+					if err != nil {
+						p.logger.Warnf("Delete(): data pruning err: %v", err)
+						return
+					}
+					// update round value for next interval calculation
+					round = currentRound
+				}
+			} else {
+				p.logger.Warnf("Delete(): unsupported interval value %v", p.config.Interval)
```
Comment (winder) [inline exporters/util/prune.go:92]: This case should never happen, kill the program?
```suggestion
				p.logger.Fatalf("Delete(): unsupported interval value %v", p.config.Interval)
```

**Review by winder — APPROVED** [review]
LGTM

**Review by Eric-Warehime — APPROVED** [review]
LGTM--I think you should remove that unnecessary line that I commented on though.

**Pull request merged**